pkg/kvstore/store/watchstore.go

```diff
@@ -0,0 +1,233 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package store
+
+import (
+	"context"
+	"strings"
+	"sync/atomic"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+)
+
+// WatchStore abstracts the operations allowing to synchronize key/value pairs
+// from a kvstore, emitting the corresponding events.
+type WatchStore interface {
+	// Watch starts watching the specified kvstore prefix, blocking until the context is closed.
+	// Depending on the implementation, it might be executed multiple times.
+	Watch(ctx context.Context, backend WatchStoreBackend, prefix string)
+
+	// NumEntries returns the number of entries synchronized from the store.
+	NumEntries() uint64
+
+	// Drain emits a deletion event for each known key. It shall be called only
+	// when no watch operation is in progress.
+	Drain()
+}
+
+// WatchStoreBackend represents the subset of kvstore.BackendOperations leveraged
+// by WatchStore implementations.
+type WatchStoreBackend interface {
+	// ListAndWatch creates a new watcher for the given prefix after listing the existing keys.
+	ListAndWatch(ctx context.Context, name, prefix string, chanSize int) *kvstore.Watcher
+}
+
+type RWSOpt func(*restartableWatchStore)
+
+// WSWithOnSyncCallback registers a function to be executed after
+// listing all keys from the kvstore for the first time. Multiple
+// callback functions can be registered.
+func RWSWithOnSyncCallback(callback func(ctx context.Context)) RWSOpt {
+	return func(rws *restartableWatchStore) {
+		rws.onSyncCallbacks = append(rws.onSyncCallbacks, callback)
+	}
+}
+
+// WSWithEntriesGauge registers a Prometheus gauge metric that is kept
+// in sync with the number of entries synchronized from the kvstore.
+func RWSWithEntriesMetric(gauge prometheus.Gauge) RWSOpt {
+	return func(rws *restartableWatchStore) {
+		rws.entriesMetric = gauge
+	}
+}
+
+type rwsEntry struct {
+	key   Key
+	stale bool
+}
+
+// restartableWatchStore implements the WatchStore interface, supporting
+// multiple executions of the Watch() operation (granted that the previous one
+// already terminated). This allows to transparently handle the case in which
+// we had to create a new etcd connection (for instance following a failure)
+// which refers to the same remote cluster.
```

Comment:
suggestion: add some discussion about the bug that this fixes, namely, that this synthesizes deletions from the local store between watch restarts.